src/reporters/service.ts

```diff
@@ -0,0 +1,302 @@
+/**
+ * Copyright (c) Microsoft Corporation.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ * http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+import fs from 'fs';
+import childProcess from 'child_process';
+import path from 'path';
+import { request } from 'https';
+import { EmptyReporter } from '../reporter';
+import { Config, Test, Suite, TestResult } from '../types';
+import { monotonicTime } from '../util';
+import storage, { BlobServiceClient, ContainerClient } from '@azure/storage-blob';
+import { formatFailure, stripAscii } from './base';
+
+export type RunResult = {
+  id: string;
+  tenantId: string;
+  workflowId: string;
+  workflowName: string;
+  workflowUrl: string;
+  repo: string;
+  branch: string;
+  triggerType: string;
+  triggerId: string;
+  triggerUrl: string;
+  numTotalTestSuites: number;
+  numFailedTestSuites: number;
+  numSkippedTestSuites: number;
+  numPassedTestSuites: number;
+  numTotalTests: number;
+  numFailedTests: number;
+  numSkippedTests: number;
+  numPassedTests: number;
+  startTime: number;
+  endTime: number;
+  status: boolean;
+}
+
+export interface TestSuiteResult {
+  path: string;
+  numTotalTests: number;
+  numFailedTests: number;
+  numSkippedTests: number;
+  numPassedTests: number;
+  duration: number;
+  tests: TestCaseResult[];
+}
+
+export interface TestCaseResult {
+  name: string;
+  status: string;
+  duration: number;
+  failureMessages: string[];
+  artifacts: string[];
```

Comment:
These seems to be local artifacts from the GHA bot. Are we doing anything special with them, or uploading using `actions/upload-artifact@v1`.

Reply:
> This looks pretty good overall. A few high-level observations:
> 
> * I think the artifacts story is not exactly clear to me.
> * Most of the information should be available through existing JSON reporter. If we only need to post-process it (which seems the case), I'd say that reading the JSON output, casting it to some TS type, and then changing it in a way we need is easier that writing a custom reporter. WDYT?
> * If we end up with a custom reporter, it should leave outside of Folio. The reporter api will be stable once Folio reaches v1.0, so there should be no risks maintaining a custom reporter in a more convenient place, instead of upstreaming it.

Hi Dmitry,

Apologies for the delayed response.

* The artifacts story basically revolves around uploading the artifacts produced by the tests to our storage to be able to correlate and show it on the custom dashboard for a particular run. For this, we upload the artifacts produced by each run onTestEnd.
* Most of the data we are consuming is available from other reporters like JSON etc, in addition we are doing upload of the artifacts and the runInfo using the hooks like onTestEnd, onBegin, onEnd etc. It seemed a cleaner way to abstract this out in a separate reporter itself since the processing happens throughout the lifetime of the run rather than sort of post-processing. We're still open to do this any other way if that would be more suitable.
* Based on the comment, do we have any existing entry point that can be leveraged for our custom reporter. If yes, could you please point us to the same. (For eg. https://jestjs.io/docs/configuration#reporters-arraymodulename--modulename-options. We have a custom reporter for Jest too through this framework.)
If not, what alternate approaches are there to have a custom reporter as part of the folio package.

Reply:
- For the artifacts, we leverage the Sas URI with Write permission for the storage account at the blob level, and this requires us to individually upload each file (as opposed to uploading the whole `test-results` folder using a container level Sas token due to security issues as there is no check on the size of files that can be uploaded using a Sas token).
- Verified using the custom reporter being provided as an argument to folio, works as expected.